cpp/src/ts/batched_kalman.cu

```diff
@@ -0,0 +1,820 @@
+//  public:
+//   int m_p = 0;
+//   int m_q = 0;
+//   int m_num_batches = 0;
+//   // double* d_ys = nullptr;
+//   double* d_Fs = nullptr;
+//   double* d_loglike = nullptr;
+//   double* d_sigma2 = nullptr;
+
+//   // memory allocators
+//   std::shared_ptr<ML::deviceAllocator> allocator;
+
+//   // TODO: This handle will probably be passed in externally later and we will only
+//   // permanently store the allocator shared_ptr.
+//   // cumlHandle m_handle;
+
+//   // // batched_jones
+//   double* d_ar = nullptr;
+//   double* d_Tar = nullptr;
+//   double* d_ma = nullptr;
+//   double* d_Tma = nullptr;
+
+//   KalmanContext(int p, int q, int num_batches, cumlHandle& handle);
+
+//   // Note: Tried to re-use these device vectors, but it caused segfaults, so we ignore them for now.
+//   // thrust::device_vector<double> d_Z_b;
+//   // thrust::device_vector<double> d_R_b;
+//   // thrust::device_vector<double> d_T_b;
+
+//   // Only allocates when the pointer is uninitialized.
+//   void allocate_if_zero(double*& ptr, size_t size);
+
+//   bool orderEquals(int p, int q, int num_batches);
+
+//   // static void resize_if_zero(thrust::device_vector<double> v, size_t size) {
+//   //   if (v.size() == 0) {
+//   //     v.resize(size);
+//   //   }
+//   // }
+
+//   ~KalmanContext() noexcept(false);
+// };
+
+// //! A global context variable which saves allocations between invocations.
+// // KalmanContext* KALMAN_CTX = nullptr;
+
+// KalmanContext::KalmanContext(int p, int q, int num_batches, cumlHandle& handle)
+//   : m_p(p), m_q(q), m_num_batches(num_batches) {
+//   // d_ys = nullptr;
+//   d_Fs = nullptr;
+//   d_loglike = nullptr;
+//   d_sigma2 = nullptr;
+//   d_ar = nullptr;
+//   d_Tar = nullptr;
+//   d_ma = nullptr;
+//   d_Tma = nullptr;
+
+//   handle.setStream(0);
+//   allocator = handle.getDeviceAllocator();
+// }
+
+// void KalmanContext::allocate_if_zero(double*& ptr, size_t size) {
+//   if (ptr == nullptr) {
+//     ptr = (double*)allocator->allocate(sizeof(double) * size, 0);
+//   }
+// }
+
+// bool KalmanContext::orderEquals(int p, int q, int num_batches) {
+//   return (m_p == p) && (m_q == q) && (m_num_batches == num_batches);
+// }
+
+// KalmanContext::~KalmanContext() noexcept(false) {
+//   ////////////////////////////////////////////////////////////
+//   // free memory
+//   // if (d_ys != nullptr) allocator->deallocate(d_ys, 0, 0);
+//   if (d_Fs != nullptr) allocator->deallocate(d_Fs, 0, 0);
+//   if (d_sigma2 != nullptr) allocator->deallocate(d_sigma2, 0, 0);
+//   if (d_loglike != nullptr) allocator->deallocate(d_loglike, 0, 0);
+
+//   if (d_ar != nullptr) allocator->deallocate(d_ar, 0, 0);
+//   if (d_Tar != nullptr) allocator->deallocate(d_Tar, 0, 0);
+//   if (d_ma != nullptr) allocator->deallocate(d_ma, 0, 0);
+//   if (d_Tma != nullptr) allocator->deallocate(d_Tma, 0, 0);
+//   CUDA_CHECK(cudaPeekAtLastError());
+// }
+
+void process_mem_usage(double& vm_usage, double& resident_set) {
```

Comment:
knitpick: this function doesn't seem to be used anywhere?

Reply:
I've removed it now. Thx.